docs/language/learn-ql/javascript/type-tracking.rst

```diff
@@ -0,0 +1,509 @@
+The ``track`` predicate maps this to a successor of that node, such as ``getDatabase()``, and
+binds ``t`` to the continuation of ``t2`` with this extra step included:
+
+.. code-block:: ql
+
+  exists(TypeTracker t2 |
+    result = firebaseDatabase(t2).track(t2, t)
+  )
+
+To understand the role of ``t`` here, note that type tracking can step *into* a property, which means
+the data flow node returned from ``track`` is not necessarily a firebase database instance, it could be
+an object *containing* a firebase database in one of its properties.
+
+For example, in the program below, the ``firebaseDatabase(t)`` predicate includes the ``obj`` node in its result,
+but with ``t`` recording the fact that the actual value being tracked is inside the ``DB`` property:
+
+.. code-block:: javascript
+
+  let obj = { DB: firebase.database() };
+  let db = obj.DB;
+
+This brings us to the last predicate. This uses ``TypeTracker::end()`` to filter out
+the paths where the firebase database instance ended up inside a property of another object,
+so it includes ``db`` but not ``obj``:
+
+.. code-block:: ql
+
+  SourceNode firebaseDatabase() {
+    result = firebaseDatabase(TypeTracker::end())
+  }
+
+Here's see an example of what this can handle now:
+
+.. code-block:: javascript
+
+  class Firebase {
+    constructor() {
+      this.db = firebase.database();
+    }
+
+    getDatabase() { return this.db; }
+
+    setForecast(value) {
+      this.getDatabase().ref("forecast").set(value); // found by firebaseSetterCall()
+    }
+  }
+
+Tracking in the whole model
+---------------------------
+We applied this pattern to ``firebaseDatabase()`` in the previous section, and it
+can just as easily apply to the other predicates.
+For reference, here's our simple Firebase model with type tracking on every predicate:
+
+.. code-block:: ql
+
+  SourceNode firebase(TypeTracker t) {
+    t.start() and
+    result = globalVarRef("firebase")
+    or
+    exists(TypeTracker t2 |
+      result = firebase(t2).track(t2, t)
+    )
+  }
+
+  SourceNode firebase() {
+    result = firebase(TypeTracker::end())
+  }
+
+  SourceNode firebaseDatabase(TypeTracker t) {
+    t.start() and
+    result = firebase().getAMethodCall("database")
+    or
+    exists(TypeTracker t2 |
+      result = firebaseDatabase(t2).track(t2, t)
+    )
+  }
+
+  SourceNode firebaseDatabase() {
+    result = firebaseDatabase(TypeTracker::end())
+  }
+
+  SourceNode firebaseRef(TypeTracker t) {
+    t.start() and
+    result = firebaseDatabase().getAMethodCall("ref")
+    or
+    exists(TypeTracker t2 |
+      result = firebaseRef(t2).track(t2, t)
+    )
+  }
+
+  SourceNode firebaseRef() {
+    result = firebaseRef(TypeTracker::end())
+  }
+
+  MethodCallNode firebaseSetterCall() {
+    result = firebaseRef().getAMethodCall("set")
+  }
+
+`Here <https://lgtm.com/query/1053770500827789481>`__ is a run of an example query using the model on one of the Firebase sample projects.
```

Comment:
Perhaps worth briefly mentioning what the example query does (it finds `firebaseSetterCall()`s).